android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt

```diff
@@ -85,12 +90,14 @@ object WorkerBinder {
   @JvmStatic
   @VisibleForTesting
   open fun bind(mappedLifecycle: Observable<WorkerEvent>, worker: Worker): WorkerUnbinder {
-    val unbindSubject = PublishRelay.create<WorkerEvent>()
-    val workerLifecycle = mappedLifecycle
-      .mergeWith(unbindSubject)
-      .takeUntil { workerEvent: WorkerEvent -> workerEvent === WorkerEvent.STOP }
+    val unbindFlow = MutableSharedFlow<WorkerEvent>(0, 1, BufferOverflow.DROP_OLDEST)
+
+    val workerLifecycle = merge(mappedLifecycle.asFlow(), unbindFlow)
+      .transformWhile { emit(it) ; it != WorkerEvent.STOP }
+      .asObservable()
+
     bindToWorkerLifecycle(workerLifecycle, worker)
```

Comment:
I prototyped a copy of bindToWorkerLifecycle that has to launch to match events, and we end up in race condtion since a subscribe on main thread is guaranteed to register before the testing code calls it. Need to look into if we should block for registering the lambda but not on the terminal event.

Reply:
More experimentation: Notice that with current Rx implementation, the issue of calling `unbinder.unbind()` on a worker thread is the same: we'll end up calling `worker.onStop` on the caller worker thread:

```kotlin
    val unbinder = PublishSubject.create<Int>()
    Observable.merge(Observable.just(1), unbinder).subscribe {
        println("$it ${Thread.currentThread().name}")
    }
    Dispatchers.Default {
        delay(100)
        unbinder.onNext(2)
    }
```
Output:
```
1 main
2 DefaultDispatcher-worker-1
```

If we use `observeOn` before `subscribe`, this fixes it, but introduces unwanted asynchronicity, hence my proposal to forbid unbinding from a worker thread.

Reply:
Regarding 1, that's what I observed as well, and was thinking we'd need to use create a coroutinecontext on the spot. We can take one in as optional in the params for testing, but the current usage does "feel" leaky since it subscribes directly, but it looks like it ends up getting coupled to the parents lifecycle through the internal logic inside it.

Regarding 2, @FranAguilera has done some research on workers binding on main thread vs background and has a proposal to make an explicit API change. WIth his proposed changes, we'd need both options.